keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+   resources to determine if the new desired ResourceRequirements fit the Node.
+   * _Case 1_: Kubelet finds new ResourceRequirements fit. It sets
+     ResizingPod.Status to true, and ResizingPod.Reason to InProgress, then
+     applies resized cgroup limits to the Pod and its Containers, and once
+     successfully done, updates Pod's ContainerStatus.ResourcesAllocated to
+     reflect the desired ResourceRequirements. It then sets ResizingPod.Status
+     to false, and ResizingPod.Reason to Success.
+     - If at the same time, a new Pod was scheduled against the capacity used by
+       this resource resize, that Pod is rejected during Kubelet admission if
+       Node has no more room.
+   * _Case 2_: Kubelet finds that desired ResourceRequirements does not fit.
+     - Kubelet checks to see if evicting lower priority Pods on the Node can
+       successfully resize the Pod. If yes, it sets ResizingPod.Status to true,
+       ResizingPod.Reason to PreEmpting, and initiates pre-emption of lower
+       priority Pods via Eviction API. Once lower priority Pods have been
+       evicted, it sets ResizingPod.Reason to InProgress, and the flow
+       continues as above.
+     - If Kubelet is unable to create enough room by evicting lower priority
+       Pods, it sets ResizingPod.Reason to FailedNodeCapacity, and
+       ResizingPod.Status to false.
+1. Scheduler, in parallel, observes that Container.ResourceRequirements and
+   ContainerStatus.ResourcesAllocated differ, and uses
+   max(ResourceRequirements, ResourcesAllocated) when computing resources
+   available on the Node.
+   * This can temporarily result in sum of Pod resources for the Node exceeding
+     Node's allocatable resources if a Pod was scheduled to that Node in parallel.
+1. Scheduler observes ResizingPod.Reason and/or ContainerStatus.ResourcesAllocated
+   fields have changed.
+   * _Case 1_: ResizingPod.Status is false, and ResizingPod.Reason is Success,
+     and Pod's Container.ResourceRequirements matches
+     ContainerStatus.ResourcesAllocated. Scheduler updates its cache with resized
+     ContainerStatus.ResourcesAllocated values.
+   * _Case 2_: ResizingPod.Status is false, and ResizingPod.Reason is
+     FailedNodeCapacity. It continues account for Node resources using
+     max(ResourceRequirements, ResourcesAllocated), and by default Kubelet is
+     expected to backoff and retry the resize at a later time, or optionally
+     when Pods depart the Node (thus creating room).
+1. The initiating actor observes that ResizingPod.Reason and/or
+   ContainerStatus.ResourcesAllocated fields have changed.
+   * _Case 1_: ResizingPod.Status is false, ResizingPod.Reason is Success,
+     and Pod's Container.ResourceRequirements matches
+     ContainerStatus.ResourcesAllocated, signifying a successful completion of
+     in-place Pod resources resizing.
+   * _Case 2_: ResizingPod.Status is false, and ResizingPod.Reason is
```

Comment:
i think this should go away.  instead, the initiating actor would see that the kubelet has been attempting to resize a pod for a period Y which exceeds its desired SLA.  as noted earlier, pods may start/stop or use less resources while a resize request is in progress.

Reply:
I hopefully clarified it this morning. During HandlePodUpdate, kubelet knows quickly if it cannot accommodate the resize (no room, no/not enough lower priority Pods to evict). In this case, it can let the initiating actor know that it's not possible. VPA can additionally have SLA where it can trigger eviction based on policy if we have been InProgress exceeding X seconds. 
cc/ @kgolab, @bskiba

Reply:
We are relying on SLA now in the updated design. Resolving this issue.